# Segmenting and Clustering Neighborhoods in Toronto

## Part 1

#### Import libraries

In [1]:
import pandas as pd
import numpy as np

#### Read the web source

In [2]:
pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')
print("Done")

Done


#### Convert the table to the dataframe

In [3]:
df_orig = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')[0]
df_orig.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

#### Delete lines with "Not assigned" in column _Borough_

In [4]:
df = df_orig[~df_orig.Borough.str.contains('Not assigned')] 
df.reset_index(drop=True, inplace=True)
df.head()

Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

#### Check if there are any multiple entries for M5A

In [5]:
df.loc[df['Postal Code'] == 'M5A']

Postal Code           Borough              Neighbourhood
2         M5A  Downtown Toronto  Regent Park, Harbourfront

Only one

#### Check if there are any entries with _Not assigned_

In [6]:
df.loc[df['Neighbourhood'] == 'Not assigned']

Empty DataFrame
Columns: [Postal Code, Borough, Neighbourhood]
Index: []

None

#### Let's see the amount of rows before and after the deletion of lines with "Not assigned" in column _Borough_

In [7]:
df_orig.shape

(180, 3)

In [8]:
df.shape

(103, 3)

The amount was decreased from 180 to 103.

## Part 2

#### Import libraries

In [9]:
import geocoder # import geocodera
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

#### Load CSV file with Canada's indexes

In [10]:
toronto_data = pd.read_csv ('Indexes_CA.csv')
toronto_data.head()

Postal Code  Latitude  Longitude
0         T0A   54.7660  -111.7174
1         T0B   53.0727  -111.5816
2         T0C   52.1431  -111.6941
3         T0E   53.6758  -115.0948
4         T0G   55.6993  -114.4529

#### Merge two dataframes by Postal code

In [11]:
df1 = df.merge(toronto_data, how='left', on=['Postal Code'])
df1.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

   Latitude  Longitude  
0   43.7545   -79.3300  
1   43.7276   -79.3148  
2   43.6555   -79.3626  
3   43.7223   -79.4504  
4   43.6641   -79.3889

## Part 3

#### Import libraries

In [12]:
import folium
from bs4 import BeautifulSoup
from sklearn.cluster import KMeans
import matplotlib.cm as cm
import matplotlib.colors as colors

#### Lets check the coordinates of Toronto for the staring point of the map visualization

In [13]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


#### For the clustering analisys only the boroughs that contain the word Toronto are going to be selected

In [14]:
df2 = df1[df1['Borough'].str.contains('Toronto',regex=False)]
df2.reset_index(drop=True, inplace=True)
df2

Postal Code           Borough  \
0          M5A  Downtown Toronto   
1          M7A  Downtown Toronto   
2          M5B  Downtown Toronto   
3          M5C  Downtown Toronto   
4          M4E      East Toronto   
5          M5E  Downtown Toronto   
6          M5G  Downtown Toronto   
7          M6G  Downtown Toronto   
8          M5H  Downtown Toronto   
9          M6H      West Toronto   
10         M5J  Downtown Toronto   
11         M6J      West Toronto   
12         M4K      East Toronto   
13         M5K  Downtown Toronto   
14         M6K      West Toronto   
15         M4L      East Toronto   
16         M5L  Downtown Toronto   
17         M4M      East Toronto   
18         M4N   Central Toronto   
19         M5N   Central Toronto   
20         M4P   Central Toronto   
21         M5P   Central Toronto   
22         M6P      West Toronto   
23         M4R   Central Toronto   
24         M5R   Central Toronto   
25         M6R      West Toronto   
26         M4S   Central Toronto   
27         M5S  Downtown Toronto   
28         M6S      West Toronto   
29         M4T   Central Toronto   
30         M5T  Downtown Toronto   
31         M4V   Central Toronto   
32         M5V  Downtown Toronto   
33         M4W  Downtown Toronto   
34         M5W  Downtown Toronto   
35         M4X  Downtown Toronto   
36         M5X  Downtown Toronto   
37         M4Y  Downtown Toronto   
38         M7Y      East Toronto   

                                        Neighbourhood  Latitude  Longitude  
0                           Regent Park, Harbourfront   43.6555   -79.3626  
1         Queen's Park, Ontario Provincial Government   43.6641   -79.3889  
2                            Garden District, Ryerson   43.6572   -79.3783  
3                                      St. James Town   43.6513   -79.3756  
4                                         The Beaches   43.6784   -79.2941  
5                                         Berczy Park   43.6456   -79.3754  
6                                  Central Bay Street   43.6564   -79.3860  
7                                            Christie   43.6683   -79.4205  
8                            Richmond, Adelaide, King   43.6496   -79.3833  
9                        Dufferin, Dovercourt Village   43.6655   -79.4378  
10  Harbourfront East, Union Station, Toronto Islands   43.6230   -79.3936  
11                           Little Portugal, Trinity   43.6480   -79.4177  
12                       The Danforth West, Riverdale   43.6803   -79.3538  
13           Toronto Dominion Centre, Design Exchange   43.6469   -79.3823  
14       Brockton, Parkdale Village, Exhibition Place   43.6383   -79.4301  
15                     India Bazaar, The Beaches West   43.6693   -79.3155  
16                     Commerce Court, Victoria Hotel   43.6492   -79.3823  
17                                    Studio District   43.6561   -79.3406  
18                                      Lawrence Park   43.7301   -79.3935  
19                                           Roselawn   43.7113   -79.4195  
20                                   Davisville North   43.7135   -79.3887  
21    Forest Hill North & West, Forest Hill Road Park   43.6966   -79.4120  
22                      High Park, The Junction South   43.6605   -79.4633  
23                  North Toronto West, Lawrence Park   43.7143   -79.4065  
24                The Annex, North Midtown, Yorkville   43.6736   -79.4035  
25                             Parkdale, Roncesvalles   43.6469   -79.4521  
26                                         Davisville   43.7020   -79.3853  
27                     University of Toronto, Harbord   43.6629   -79.3987  
28                                 Runnymede, Swansea   43.6512   -79.4828  
29                        Moore Park, Summerhill East   43.6899   -79.3853  
30          Kensington Market, Chinatown, Grange Park   43.6541   -79.3978  
31  Summerhill West, Rathnelly, South Hill, Forest...   43.6861   -79.4025  
32  CN Tower, King an

#### Lets visualize these boroughs on the map

In [15]:
map_toronto = folium.Map(location=[latitude, longitude],zoom_start=10)

for lat,lng,borough,neighbourhood in zip(df2['Latitude'],df2['Longitude'],df2['Borough'],df2['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
    [lat,lng],
    radius=5,
    popup=label,
    color='blue',
    fill=True,
    fill_color='#3186cc',
    fill_opacity=0.7,
    parse_html=False).add_to(map_toronto)
map_toronto



####  Using Kmeans 5 clusters is assigned and the column with the Cluster Name is added to the data frame

In [16]:
kclusters=5
toronto_clusters = df2.drop(['Postal Code','Borough','Neighbourhood'],1)
kmeans = KMeans(n_clusters = kclusters,random_state=0).fit(toronto_clusters)
kmeans.labels_
df2.insert(0, 'Cluster Names', kmeans.labels_)

In [17]:
df2.head()

Cluster Names Postal Code           Borough  \
0              4         M5A  Downtown Toronto   
1              4         M7A  Downtown Toronto   
2              4         M5B  Downtown Toronto   
3              4         M5C  Downtown Toronto   
4              2         M4E      East Toronto   

                                 Neighbourhood  Latitude  Longitude  
0                    Regent Park, Harbourfront   43.6555   -79.3626  
1  Queen's Park, Ontario Provincial Government   43.6641   -79.3889  
2                     Garden District, Ryerson   43.6572   -79.3783  
3                               St. James Town   43.6513   -79.3756  
4                                  The Beaches   43.6784   -79.2941

#### The 5 clusters on the map of Toronto

In [18]:
# create map
map_clusters = folium.Map(location=[latitude, longitude],zoom_start=10)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, neighbourhood, cluster in zip(df2['Latitude'], df2['Longitude'], df2['Neighbourhood'], df2['Cluster Names']):
    label = folium.Popup(' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters